analyzers/tests/SonarAnalyzer.Test/Rules/MethodsShouldNotHaveIdenticalImplementationsTest.cs

```diff
@@ -22,36 +22,151 @@
 using CS = SonarAnalyzer.Rules.CSharp;
 using VB = SonarAnalyzer.Rules.VisualBasic;
 
-namespace SonarAnalyzer.Test.Rules
+namespace SonarAnalyzer.Test.Rules;
+
+[TestClass]
+public class MethodsShouldNotHaveIdenticalImplementationsTest
 {
-    [TestClass]
-    public class MethodsShouldNotHaveIdenticalImplementationsTest
-    {
-        private readonly VerifierBuilder builderCS = new VerifierBuilder<CS.MethodsShouldNotHaveIdenticalImplementations>();
-        private readonly VerifierBuilder builderVB = new VerifierBuilder<VB.MethodsShouldNotHaveIdenticalImplementations>();
+    private readonly VerifierBuilder builderCS = new VerifierBuilder<CS.MethodsShouldNotHaveIdenticalImplementations>();
+    private readonly VerifierBuilder builderVB = new VerifierBuilder<VB.MethodsShouldNotHaveIdenticalImplementations>();
+
+    [TestMethod]
+    public void MethodsShouldNotHaveIdenticalImplementations() =>
+        builderCS.AddPaths("MethodsShouldNotHaveIdenticalImplementations.cs").WithOptions(ParseOptionsHelper.FromCSharp8).Verify();
+
+    [DataTestMethod]
+    [DataRow("", "")]
+    [DataRow("where T: struct", "where T: struct")]
+    [DataRow("where T: class", "where T: class")]
+    [DataRow("where T: unmanaged", "where T: unmanaged")]
+    [DataRow("where T: new()", "where T: new()")]
+    [DataRow("where T: IEquatable<T>, IComparable",  "where T: System.IComparable, IEquatable<T>")]
+    public void MethodsShouldNotHaveIdenticalImplementations_MethodTypeParameters_NonCompliant(string constraint1, string constraint2) =>
+        builderCS.AddSnippet($$"""
+            using System;
+            public static class TypeConstraints
+            {
+                public static bool Compare1<T>(T? value1, T value2) {{constraint1}} // Secondary
+                {
+                    Console.WriteLine(value1);
+                    Console.WriteLine(value2);
+                    return true;
+                }
+
+                public static bool Compare2<T>(T? value1, T value2) {{constraint2}} // Noncompliant
+                {
+                    Console.WriteLine(value1);
+                    Console.WriteLine(value2);
+                    return true;
+                }
+            }
+            """).WithOptions(ParseOptionsHelper.FromCSharp9).Verify();
+
+    [CombinatorialData]
+    public void MethodsShouldNotHaveIdenticalImplementations_MethodTypeParameters_Compliant(
+        [DataValues("where T: struct", "where T: class")] string constraint1,
+        [DataValues("", "where T: unmanaged", "where T: new()")] string constraint2) =>
```

Comment:
This test the combinations
"where T: struct", ""
"where T: class", ""
"where T: struct", "where T: unmanaged"
"where T: class", "where T: unmanaged"

and so on

Reply:
Why is that interesting to test?

Reply:
Because our coverage tool is not able to correctly identify that we miss a lot of combinations that are untested here:
https://github.com/SonarSource/sonar-dotnet/pull/9398/files#diff-04468862a10b44614d29645ca93eff4ff27fa55b59e226ecb2613c286df6b784R104-R107

This test adds more coverage for these conditions.

Reply:
I see. But then it would be better to test all combinations.

Reply:
See 3c35dac7888ad0cba90d06d5e6422c98f4f86c3b